### PR TITLE
ICU-13743 Adding number permutation test.

### PR DESCRIPTION
**Pull request by sffc**
Adds a test suite in C++ and Java to test many permutations of options in NumberFormatter.

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-13743
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/test/intltest/numbertest_permutation.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+  int32_t lineNumber = 1;
+  int32_t lineLength;
+  for (const auto& actualLine : results) {
+    const UChar* lineBuf = ucbuf_readline(f.getAlias(), &lineLength, status);
+    if (lineBuf == nullptr) {
+      errln("More lines generated than are in the data file!");
+      break;
+    }
+    UnicodeString expectedLine(lineBuf, lineLength - 1);
```
Comment (markusicu) [inline icu4c/source/test/intltest/numbertest_permutation.cpp]: How about skipping comment lines, and empty lines following comments or other empties?
For example, I think we should prepend a copyright header.

File: icu4c/source/test/intltest/numbertest_permutation.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    // Decimal Separator Display
+    u"decimal-always",
+    0,
+};
+
+static const double kNumbersToTest[] {
+    0,
+    91827.3645,
+    0.22222
+};
```
Comment (markusicu) [inline icu4c/source/test/intltest/numbertest_permutation.cpp]: Optional: add a negative value? Or make one of these negative?

File: icu4c/source/test/intltest/numbertest.h [hunk 3]
```diff
@@ -287,6 +287,13 @@ class NumberRangeFormatterTest : public IntlTestWithFieldPosition {
       const char16_t* expected);
 };
 
+class NumberPermutationTest : public IntlTest {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/test/intltest/numbertest.h]: Acknowledged

File: icu4c/source/test/intltest/numbertest_permutation.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+#include <vector>
+#include <iostream>
+#include <fstream>
+
+#include "numbertest.h"
+#include "ucbuf.h"
+#include "unicode/numberformatter.h"
+
+void NumberPermutationTest::runIndexedTest(int32_t index, UBool exec, const char*& name, char*) {
+  if (exec) {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_permutation.cpp]: I re-formatted the file using clang-format.

File: icu4c/source/test/intltest/numbertest_permutation.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  TESTCASE_AUTO_BEGIN;
+    TESTCASE_AUTO(testPermutations);
+  TESTCASE_AUTO_END;
+}
+
+static const char16_t* kSkeletonParts[] = {
+    // Notation
+    u"compact-short",
+    u"scientific/+ee/sign-always",
+    0,
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_permutation.cpp]: Done

File: icu4c/source/test/intltest/numbertest_permutation.cpp (hunk 2 above)
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_permutation.cpp]: Done

File: icu4c/source/test/intltest/numbertest_permutation.cpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+ * To regenerate that file, run intltest with the -G option.
+ */
+void NumberPermutationTest::testPermutations() {
+  IcuTestErrorCode status(*this, "testPermutations");
+
+  const struct LocaleData {
+    Locale locale;
+    const char16_t* ustring;
+  } localesToTest[] = {
+    { "es-MX", u"es-MX" },
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_permutation.cpp]: I didn't want to put more logic into the loop than necessary.  The Locale class doesn't have the greatest reputation for performance, so I wanted to pre-compute it.

File: icu4c/source/test/intltest/numbertest_permutation.cpp [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    { "zh-TW", u"zh-TW" },
+    { "bn-BD", u"bn-BD" },
+  };
+
+  // Convert kSkeletonParts to a more convenient data structure
+  auto skeletonParts = std::vector<std::vector<const char16_t *>>();
+  auto currentSection = std::vector<const char16_t *>();
+  for (int32_t i = 0; i < UPRV_LENGTHOF(kSkeletonParts); i++) {
+    const char16_t *skeletonPart = kSkeletonParts[i];
+    if (skeletonPart == 0) {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_permutation.cpp]: Done

File: icu4c/source/test/intltest/numbertest_permutation.cpp (hunk 1 above)
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_permutation.cpp]: The design of this test is to treat the file as a golden file, not intended to be edited by humans except when using the `-G` flag to intltest.

I added a comment block.

**Comment by sffc** [comment]
OK, the Java version is added.  The test did its job and exposed one subtle bug in the Java implementation, which I fixed.  The rest of the results indicate that the C and J implementations are very compatible in behavior.

Would you like this test to run always, or only in exhaustive mode?  It takes 403 ms in C++ and 817 ms in Java.  I think it's slow in part because the formatters each run exactly 3 times, which is the least-optimal number.  I have a ticket to do some additional profiling.

**Comment by markusicu** [comment]
Re test runtime: I would always run at least part of the test. Are your runtimes from a debug build or an optimized build? If it seems slow, and it's not an exhaustive test, you could do something like do only one of the locales for each test case (rotate through them).

**Comment by markusicu** [comment]
Please fix/update the PR description text.

File: icu4c/source/test/testdata/numberpermutationtest.txt [hunk 8]
```diff
@@ -1,9100 +1,9103 @@
+# © 2017 and later: Unicode, Inc. and others.
```
Comment (markusicu) [inline icu4c/source/test/testdata/numberpermutationtest.txt]: 2019 and later

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberPermutationTest.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                    break;
+                }
+                assertEquals("Line #" + lineNumber + " differs",  //
+                    expectedLine, actualLine);
+                lineNumber++;
+            }
+            if (f.readLine() != null) {
+                errln("Fewer lines generated than are in the data file!");
+            }
+        } catch (IOException err) {
```
Comment (markusicu) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberPermutationTest.java]: Optional: Just have your test function throws IOException.

File: icu4c/source/test/testdata/numberpermutationtest.txt (hunk 8 above)
Comment (sffc) [reply icu4c/source/test/testdata/numberpermutationtest.txt]: Done

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberPermutationTest.java (hunk 9 above)
Comment (sffc) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberPermutationTest.java]: Done

**Comment by sffc** [comment]
For non-exhaustive mode, I made it stop at 1000 lines.

I updated the PR description.

**Comment by sffc** [comment]
Squashed.  Note that I added one more commit before I squashed, d7d4e72, slightly changing the quick mode condition.

**Pull request merged**